Fix #4449: Modify the design of the chapter list in lessons tab

JishnuGoyal opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
  
###Quick before an after preview:
  
![image](https://user-images.githubusercontent.com/64526117/215283148-ed1ba94a-304b-4a24-bd7a-fdde85ddeabe.png)


This PR, as part of the GSoC project: Interactive Onboarding Flow, modifies the look of the recycler view adapter items on the topic lessons tab.
Now there are different looking item view holders for completed, in-progress, locked and not-started chapter which results in a clear CTA for the users, and they are more likely to click on the correct option. Link to the UI mock: https://xd.adobe.com/view/3dca36c2-5115-419c-b25e-0f10526b077c-6899/screen/558af437-bec4-4719-8d30-291da10bbc9d/specs/

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

Before:
![image](https://user-images.githubusercontent.com/64526117/186780683-893df4ec-bff3-4116-aa06-d17e78742d19.png)

![image](https://user-images.githubusercontent.com/64526117/186780700-ec6996b0-6655-48db-8a99-3f9dce44dd83.png)


After:
![image](https://user-images.githubusercontent.com/64526117/186780410-471d9336-7234-4b4c-8598-e569a1a6c751.png)

![image](https://user-images.githubusercontent.com/64526117/186780443-8ad90c76-92fd-4fd0-a967-4139031ee9bb.png)

![image](https://user-images.githubusercontent.com/64526117/186780538-27861097-06fe-4463-b678-7c56484106ed.png)

![image](https://user-images.githubusercontent.com/64526117/186780591-719eeb95-e7e4-412a-9f62-eb4e160a28cb.png)

Tablet View: (Pixel C API 30)
Portrait
![image](https://user-images.githubusercontent.com/64526117/186957483-b22b158b-8f1d-4aca-9d23-fb846013dba7.png)
![image](https://user-images.githubusercontent.com/64526117/187538798-577115cb-d80f-401b-93e1-2aca0194cf79.png)

Landscape
![image](https://user-images.githubusercontent.com/64526117/187538383-07742353-46e0-42ac-b10c-3418298bc34c.png)

with green border:
![image](https://user-images.githubusercontent.com/64526117/187538715-a7ca9e45-1248-49cc-98da-42eb82de99d8.png)

Accessibility scanner output:
The background to foreground ratio of these view have been kept this way on-purpose to let the users know that those chapters are locked.
![image](https://user-images.githubusercontent.com/64526117/187724589-e25d78dd-1a49-4afc-ad36-d6694916bcc5.png)

![image](https://user-images.githubusercontent.com/64526117/187724775-26ebed57-33b4-4a0d-b398-e744a4d0db12.png)

RTL:
![image](https://user-images.githubusercontent.com/64526117/190875380-9b54cf13-187c-4876-af64-450bb5e75dfa.png)

![image](https://user-images.githubusercontent.com/64526117/190875387-0fd18eb1-dd74-4943-81a2-96ae361dacff.png)

Talkback:


https://user-images.githubusercontent.com/64526117/190875517-de736464-31a5-43e3-af6d-89cd04f77376.mp4

JishnuGoyal committed:
add locked lesson view, modify completed lesson view

JishnuGoyal committed:
add borders, not started chapter view, recycler multi adapter

JishnuGoyal committed:
add divider item decoration

JishnuGoyal committed:
commit before checkout -- all design modified

JishnuGoyal committed:
add divider item decoration grey color

JishnuGoyal committed:
modify chapter recycler items to have 0 margins

JishnuGoyal committed:
remove unused lines

JishnuGoyal committed:
add in-progress chapter view item

JishnuGoyal committed:
use frame layout to make dotted line thinner

JishnuGoyal committed:
lint

JishnuGoyal committed:
remove unused resources and code analysis

JishnuGoyal committed:
lint

JishnuGoyal committed:
merge with develop + resolve conflicts

JishnuGoyal committed:
rename file to more describing name

JishnuGoyal commented:
Please take an initial pass at this
PTAL @BenHenning

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentPresenter.kt:
```diff
@@ -227,12 +230,47 @@ class TopicLessonsFragmentPresenter @Inject constructor(
+      .build()
+  }
+
+  enum class ChapterViewType {
```
We should add KDocs for the enum & its constants since it's part of the class's public API.

JishnuGoyal replied:
this should be private instead, adding a private modifier. Please let me know if a helpful comment is still needed, though im sure that the enums are self explanatory.

BenHenning replied:
In that case, can we add the private modifier? I don't think any additional documentation will be needed for the private context.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentPresenter.kt:
```diff
@@ -227,12 +230,47 @@ class TopicLessonsFragmentPresenter @Inject constructor(
+          ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES -> ChapterViewType.CHAPTER_LOCKED
+          ChapterPlayState.COMPLETED -> ChapterViewType.CHAPTER_COMPLETED
+          ChapterPlayState.IN_PROGRESS_SAVED -> ChapterViewType.CHAPTER_IN_PROGRESS
+          else -> ChapterViewType.CHAPTER_NOT_STARTED
```
I suggest adding the other constants rather than using 'else' to create an exhaustive ``when`` that will break if future play states are added (thus forcing the author to consider what to do here explicitly).

JishnuGoyal replied:
done. Could you please cross check if the corresponding view states seem correct?

BenHenning replied:
Done--everything seems to look correct to me.

BenHenning commented on app/src/main/res/drawable/border_white.xml:
```diff
@@ -0,0 +1,6 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+  android:shape="rectangle">
+  <stroke android:color="@color/color_def_bright_green" android:width="4dp"/>
+  <solid android:color="@color/color_def_white"/>
```
This doesn't seem to match the drawable name--ditto for the other one (border_bright_green). Could these drawable names perhaps be clearer?

JishnuGoyal replied:
done -- pls let me know if the updated names seem correct or should be renamed.

BenHenning commented on app/src/main/res/drawable/divider_item_decoration_grey.xml:
```diff
@@ -0,0 +1,9 @@
+      <size android:height="1dp" />
+    </shape>
+  </item>
+</selector>
```
Please add EOF newline.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/drawable/divider_item_decoration_grey.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
```
Is a selector actually needed here? It seems like the same shape is always used.

Also regarding name, perhaps: ``grey_recycler_view_item_decoration_divider`` might be clearer?

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/drawable/ic_baseline_lock_24.xml:
```diff
@@ -1,10 +1,5 @@
-  <path
-    android:fillColor="#FFFFFF"
-    android:pathData="M18,8h-1L17,6c0,-2.76 -2.24,-5 -5,-5S7,3.24 7,6v2L6,8c-1.1,0 -2,0.9 -2,2v10c0,1.1 0.9,2 2,2h12c1.1,0 2,-0.9 2,-2L20,10c0,-1.1 -0.9,-2 -2,-2zM12,17c-1.1,0 -2,-0.9 -2,-2s0.9,-2 2,-2 2,0.9 2,2 -0.9,2 -2,2zM15.1,8L8.9,8L8.9,6c0,-1.71 1.39,-3.1 3.1,-3.1 1.71,0 3.1,1.39 3.1,3.1v2z" />
+<vector android:autoMirrored="true" android:height="24dp"
```
What changed between this & the previous one? It seems like dimensions & formatting--anything else?

JishnuGoyal replied:
deleted the other one

BenHenning commented on app/src/main/res/drawable/ic_baseline_lock_20.xml:
```diff
@@ -0,0 +1,5 @@
+<vector android:autoMirrored="true" android:height="20dp"
```
Why is this needed when we have ic_baseline_lock_24? What's the difference between size? If it's size, why not just use the other one and scale it down (since it's a vector graphic, anyway)?

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/lessons_completed_chapter_view.xml:
```diff
@@ -14,51 +13,55 @@
-      android:layout_width="wrap_content"
-      android:layout_height="wrap_content"
-      android:layout_marginStart="4dp"
+      android:layout_width="60dp"
```
Does this need to be a dimen? How does the layout look on tablets?

JishnuGoyal replied:
Dimen might not be required. Added the tablet screenshots to the PR description, please let me know if dimens are required.

BenHenning commented on app/src/main/res/layout/lessons_completed_chapter_view.xml:
```diff
@@ -14,51 +13,55 @@
-      android:layout_marginStart="4dp"
+      android:layout_width="60dp"
+      android:layout_height="match_parent"
+      android:layout_marginStart="0dp"
```
Can be removed.

JishnuGoyal replied:
not sure -- which line is suggested to be removed?

BenHenning replied:
The margin of 0dp--a 0dp margin doesn't actually change anything unless you're overwriting a style.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/lessons_completed_chapter_view.xml:
```diff
@@ -14,51 +13,55 @@
       android:fontFamily="sans-serif"
+      android:gravity="center"
       android:importantForAccessibility="@{viewModel.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? View.IMPORTANT_FOR_ACCESSIBILITY_YES : View.IMPORTANT_FOR_ACCESSIBILITY_NO}"
+      android:paddingStart="8dp"
+      android:paddingTop="12dp"
+      android:paddingEnd="8dp"
+      android:paddingBottom="12dp"
```
Ditto for these--they may need to be dimensions so that they can be correctly customized for tablets.

JishnuGoyal replied:
Please check the newly added screenshots for tablet view and let me know if dimens are still required

BenHenning replied:
Can you include landscape screenshots for the case that includes the dark green border?

Also, do you know why there's a bit of extra above/below whitespace in the tablet landscape case? That seems like a visual bug.

BenHenning commented on app/src/main/res/layout/lessons_completed_chapter_view.xml:
```diff
@@ -14,51 +13,55 @@
       android:textSize="14sp" />
   </LinearLayout>
+
+    <View
```
The indentation looks incorrect here--please fix.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/lessons_completed_chapter_view.xml:
```diff
@@ -14,51 +13,55 @@
     android:layout_width="match_parent"
     android:layout_height="wrap_content"
+    android:orientation="vertical">
+  <LinearLayout
```
Rather than nesting LinearLayouts, could we just use a single constraint layout?

JishnuGoyal replied:
this layout file also contains the recycler view divider decoration at the bottom, which seems easier to do with two nested linearlayouts rather than constraint layout. Would it be okay to let it be this way or is there another aspect to why we must change this?

BenHenning replied:
Is it much more complicated with ConstraintLayout? It doesn't seem like it would be, but maybe I'm missing something? Generally, ConstraintLayout is preferred over nested layouts since it avoids multiple layout/measure passes (though the benefit is more noticeable when nesting relative layouts than linear).

JishnuGoyal replied:
I've added constraint layout to the file you mentioned + one other file that has a similar xml. The other two files have a relatively complex view hierarchy and I feel they are better with the linear layout (they are actually a bit problematic to switch to the constraint layout as the nested linear layout itself plays a role). Please let me know what you think.

BenHenning commented on app/src/main/res/layout/topic_lessons_story_summary.xml:
```diff
@@ -138,37 +138,36 @@
       </LinearLayout>
 
-      <LinearLayout
+      <FrameLayout
```
It may be worth adding a comment here explaining why a FrameLayout is needed (since it's unusual to see a case where we are intentionally using overdraw to fix built-in drawable padding).

JishnuGoyal replied:
done

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentTest.kt:
```diff
@@ -311,6 +311,37 @@ class TopicLessonsFragmentTest {
   }
 
+  @Test
+  fun testLessonsPlayFragment_loadRatiosTopic_chapterNotStartedIsCorrectlyDisplayed() {
```
All of the different states now possible should be tested, not just the 'not started' state.

JishnuGoyal replied:
done

BenHenning reviewed (commented):
Thanks @JishnuGoyal! Took a first pass. I think after these comments are addressed you can send the PR to Rajat for codeowners review.

JishnuGoyal committed:
add tests and correct some errors

JishnuGoyal committed:
suggested changes + fix some problems

JishnuGoyal committed:
rename resources

JishnuGoyal committed:
lint

JishnuGoyal committed:
lint

JishnuGoyal commented:
Thanks @BenHenning 
Please take another pass on this PR as I think there are a number of things that require another check.
@rt4914 would it be possible for you to PTAL for codeowners?

JishnuGoyal committed:
add missing eof

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentPresenter.kt:
```diff
@@ -236,7 +236,11 @@ class TopicLessonsFragmentPresenter @Inject constructor(
           ChapterPlayState.COMPLETED -> ChapterViewType.CHAPTER_COMPLETED
           ChapterPlayState.IN_PROGRESS_SAVED -> ChapterViewType.CHAPTER_IN_PROGRESS
-          else -> ChapterViewType.CHAPTER_NOT_STARTED
+          ChapterPlayState.IN_PROGRESS_NOT_SAVED -> ChapterViewType.CHAPTER_IN_PROGRESS
```
You can combine these together, e.g.:

```kotlin
ChapterPlayState.IN_PROGRESS_SAVED, ChapterPlayStage.IN_PROGRESS_NOT_SAVED ->
  ChapterViewType.CHAPTER_IN_PROGRESS
```

Ditto for the others (where relevant).

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/topic_lessons_story_summary.xml:
```diff
@@ -144,7 +144,8 @@
       </LinearLayout>
-
+      <!-- FrameLayout is used here instead of LinearLayout to properly adjust the visible thickness
+       of the dashed divider-->
```
```suggestion
       of the dashed divider by allowing the divider to render slightly on top of the container. -->
```

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/topic_lessons_story_summary.xml:
```diff
@@ -144,7 +144,8 @@
         </LinearLayout>
       </LinearLayout>
-
+      <!-- FrameLayout is used here instead of LinearLayout to properly adjust the visible thickness
```
```suggestion
      <!-- FrameLayout is used here instead of LinearLayout to properly adjust the spacing
```

Along with my suggestion below, may be worth moving words up/down per line-wrapping.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/lessons_not_started_chapter_view.xml:
```diff
@@ -13,26 +13,25 @@
         android:layout_height="match_parent"
         android:layout_marginStart="0dp"
-        android:background="@drawable/border_bright_green"
+        android:background="@drawable/chapter_dark_green_bright_green_border_background"
```
```suggestion
        android:background="@drawable/chapter_dark_green_fill_bright_green_border_background"
```

WDYT about this name, instead, or something similar? The double color in a row part of the name is slightly confusing at the moment. Ditto for the others.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/lessons_not_started_chapter_view.xml:
```diff
@@ -13,26 +13,25 @@
       android:layout_width="match_parent"
       android:layout_height="wrap_content"
-      android:background="@drawable/border_white"
-      android:clickable="@{viewModel.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? true : false}"
```
Is this behavior change for accessibility reasons?

JishnuGoyal replied:
the clickability of locked lessons view is always false. In the earlier version we had conditional clickability as only one layout res was used, but now we have multiple.

BenHenning reviewed (commented):
Thanks @JishnuGoyal. Took another pass--PTAL.

rt4914 commented on app/src/main/res/layout/lessons_completed_chapter_view.xml:
```diff
@@ -0,0 +1,67 @@
+
+      <TextView
+        android:id="@+id/chapter_index"
+        android:layout_width="60dp"
```
We should refrain from giving fixed widths to `TextView's` .

Suggest making this `wrap_content` and using `minWidth=60dp` instead.

JishnuGoyal replied:
done

rt4914 commented on app/src/main/res/layout/lessons_in_progress_chapter_view.xml:
```diff
@@ -0,0 +1,73 @@
+      <TextView
+        android:id="@+id/chapter_index"
+        android:layout_width="20dp"
+        android:layout_height="20dp"
```
Ditto.
Refrain from giving fixed width/height.

Use `minWidth/Height` instead

JishnuGoyal replied:
done

rt4914 commented on app/src/main/res/layout/lessons_locked_chapter_view.xml:
```diff
@@ -0,0 +1,74 @@
+      <TextView
+        android:id="@+id/chapter_index"
+        android:layout_width="20dp"
+        android:layout_height="20dp"
```
Ditto

JishnuGoyal replied:
done

rt4914 commented on app/src/main/res/layout/lessons_not_started_chapter_view.xml:
```diff
@@ -0,0 +1,64 @@
+
+      <TextView
+        android:id="@+id/chapter_index"
+        android:layout_width="60dp"
```
Ditto

JishnuGoyal replied:
done

rt4914 commented on app/src/main/res/layout/topic_lessons_story_summary.xml:
```diff
@@ -138,37 +138,37 @@
-
-      <LinearLayout
+      <!-- FrameLayout is used here instead of LinearLayout to properly adjust the visible thickness
+       of the dashed divider-->
```
Missing space before `-->`

JishnuGoyal replied:
done

rt4914 reviewed (changes requested):
@JishnuGoyal Can you please share mock links as they are important to verify the UI.

Also please add following screenshots:

1. Accessibility Scanner Output
2. RTL Screenshot

Also, doers this new design contain Dark-mode mocks too?

JishnuGoyal committed:
add private identifier to enum

JishnuGoyal committed:
remove 0dp margin

JishnuGoyal committed:
commit requested changes

JishnuGoyal committed:
add framelayout to sw600dp-land for tablets + lint fix for file

JishnuGoyal committed:
rename resources

JishnuGoyal committed:
lint

JishnuGoyal committed:
switch to constraint layout

JishnuGoyal commented:
PTAL @BenHenning

BenHenning reviewed (commented):
Thanks @JishnuGoyal! Everything here LGTM except the failing CI checks--please fix and send back.

rt4914 commented:
@JishnuGoyal PTAL at this comment https://github.com/oppia/oppia-android/pull/4535#pullrequestreview-1089345084

JishnuGoyal committed:
remove the "period sign" in chapter index.

JishnuGoyal commented:
Thanks @rt4914 .

> Can you please share mock links as they are important to verify the UI.
Also please add following screenshots:
Accessibility Scanner Output
RTL Screenshot
Also, doers this new design contain Dark-mode mocks too?

Mock links have been added to the PR description. Dark mode is not available in the mocks.
I have added accessibility scanner outputs, and the reason along with why we should go with that font ragardless of the contrast issue. (Please let me know if this still needs changes so that I could start a discussion on an email thread with Sean and Chantel and loop you in)
Also added the RTL screenshots. PTAL @rt4914

JishnuGoyal commented:
This PR is almost complete, but one functionality to have the lessons list expanded is still left which will be added soon. I'll  request @BenHenning to merge this after that functionality is successfully added along with its tests.

JishnuGoyal committed:
Merge branch 'develop' into modify_chapter_list_design

JishnuGoyal committed:
add field to topicsummary proto, and complete the func for opening chapter list by default

JishnuGoyal committed:
lint

JishnuGoyal commented:
Hi, @BenHenning. The tests seem to fail without an error -- PTAL

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -242,6 +242,9 @@ message TopicSummary {
   // Specifics about whether this topic is playable.
   TopicPlayAvailability topic_play_availability = 6;
+
+  // The ID of the first story of this topic
```
```suggestion
  // The ID of the first story of this topic.
```

JishnuGoyal replied:
done

BenHenning reviewed (commented):
Thanks @JishnuGoyal. Took another pass. Waiting for the PR to be finished & all tests to be passing before approving.

> Hi, @BenHenning. The tests seem to fail without an error -- PTAL

Checking app-shard8, I see:

```
//app:src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest (cached) PASSED in 30.6s
//app:src/sharedTest/java/org/oppia/android/app/devoptions/ViewEventLogsActivityTest (cached) PASSED in 23.1s
//app:src/sharedTest/java/org/oppia/android/app/help/HelpFragmentTest (cached) PASSED in 58.2s
//app:src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest (cached) PASSED in 23.8s
//app:src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingActivityTest (cached) PASSED in 9.5s
//app:src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonActivityTest (cached) PASSED in 25.0s
//app:src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditFragmentTest (cached) PASSED in 33.4s
//app:src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileListFragmentTest (cached) PASSED in 28.8s
//app:src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest (cached) PASSED in 56.1s
//app:src/sharedTest/java/org/oppia/android/app/topic/info/TopicInfoFragmentTest (cached) PASSED in 29.1s
//app:src/sharedTest/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivityTest (cached) PASSED in 50.2s
//app:src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest (cached) PASSED in 219.2s
//app:src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCountTest (cached) PASSED in 18.5s
//app:src/test/java/org/oppia/android/app/topic/info/TopicInfoFragmentLocalTest (cached) PASSED in 23.3s
//app:src/sharedTest/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentTest FAILED in 37.6s
```

(So we know that TopicLessonsFragmentTest failed this shard and should be run locally to investigate).

Scrolling up a bit further in the raw logs shows:

```
2022-09-03T20:18:09.6609612Z Time: 36.504
2022-09-03T20:18:09.6609812Z There was 1 failure:
2022-09-03T20:18:09.6610302Z 1) testLessonsPlayFragment_loadFractionsTopic_clickChap_chapterMarkedAsNotStarted_opensExpAct(org.oppia.android.app.topic.lessons.TopicLessonsFragmentTest)
2022-09-03T20:18:09.6610978Z androidx.test.espresso.base.DefaultFailureHandler$AssertionFailedWithCauseError: Wanted to match 1 intents. Actually matched 0 intents.
2022-09-03T20:18:09.6611291Z 
2022-09-03T20:18:09.6613080Z IntentMatcher: (has extras: has bundle with: key: is "ExplorationActivity.exploration_id" value: is "umPkwp0L1M0-" and has extras: has bundle with: key: is "ExplorationActivity.story_id" value: is "wANbh4oOClga" and has extras: has bundle with: key: is "ExplorationActivity.topic_id" value: is "GJ2rLXRKD5hw" and has extras: has bundle with: key: is "ExplorationActivity.profile_id" value: is <0> and has extras: has bundle with: key: is "ExplorationActivity.is_checkpointing_enabled_key" value: is <true> and has extras: has bundle with: key: is "ExplorationActivity.backflow_screen" value: is <0> and has component: has component with: class name: is "org.oppia.android.app.player.exploration.ExplorationActivity" package name: an instance of java.lang.String short class name: an instance of java.lang.String)
2022-09-03T20:18:09.6614218Z 
2022-09-03T20:18:09.6614307Z Matched intents:[]
2022-09-03T20:18:09.6614433Z 
2022-09-03T20:18:09.6614519Z Recorded intents:[]
2022-09-03T20:18:09.6614792Z 	at java.base/java.lang.Thread.getStackTrace(Thread.java:1654)
2022-09-03T20:18:09.6622398Z 	at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:96)
2022-09-03T20:18:09.6623014Z 	at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:59)
2022-09-03T20:18:09.6623602Z 	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:324)
2022-09-03T20:18:09.6624138Z 	at androidx.test.espresso.ViewInteraction.check(ViewInteraction.java:306)
2022-09-03T20:18:09.6624561Z 	at androidx.test.espresso.intent.Intents.intended(Intents.java:189)
2022-09-03T20:18:09.6624970Z 	at androidx.test.espresso.intent.Intents.intended(Intents.java:170)
2022-09-03T20:18:09.6625660Z 	at org.oppia.android.app.topic.lessons.TopicLessonsFragmentTest.testLessonsPlayFragment_loadFractionsTopic_clickChap_chapterMarkedAsNotStarted_opensExpAct(TopicLessonsFragmentTest_updated.kt:613)
2022-09-03T20:18:09.6626222Z 	... 19 trimmed
2022-09-03T20:18:09.6626892Z Caused by: junit.framework.AssertionFailedError: Wanted to match 1 intents. Actually matched 0 intents.
2022-09-03T20:18:09.6627144Z 
2022-09-03T20:18:09.6721361Z IntentMatcher: (has extras: has bundle with: key: is "ExplorationActivity.exploration_id" value: is "umPkwp0L1M0-" and has extras: has bundle with: key: is "ExplorationActivity.story_id" value: is "wANbh4oOClga" and has extras: has bundle with: key: is "ExplorationActivity.topic_id" value: is "GJ2rLXRKD5hw" and has extras: has bundle with: key: is "ExplorationActivity.profile_id" value: is <0> and has extras: has bundle with: key: is "ExplorationActivity.is_checkpointing_enabled_key" value: is <true> and has extras: has bundle with: key: is "ExplorationActivity.backflow_screen" value: is <0> and has component: has component with: class name: is "org.oppia.android.app.player.exploration.ExplorationActivity" package name: an instance of java.lang.String short class name: an instance of java.lang.String)
2022-09-03T20:18:09.6722514Z 
2022-09-03T20:18:09.6722609Z Matched intents:[]
2022-09-03T20:18:09.6722907Z 
2022-09-03T20:18:09.6722999Z Recorded intents:[]
2022-09-03T20:18:09.6723266Z 	at junit.framework.Assert.fail(Assert.java:50)
2022-09-03T20:18:09.6723815Z 	at androidx.test.espresso.intent.VerificationModes$Times.verify(VerificationModes.java:80)
2022-09-03T20:18:09.6724291Z 	at androidx.test.espresso.intent.Intents.internalIntended(Intents.java:346)
2022-09-03T20:18:09.6724714Z 	at androidx.test.espresso.intent.Intents$2.check(Intents.java:193)
2022-09-03T20:18:09.6725169Z 	at androidx.test.espresso.ViewInteraction$SingleExecutionViewAssertion.check(ViewInteraction.java:425)
2022-09-03T20:18:09.6725804Z 	at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:288)
2022-09-03T20:18:09.6726204Z 	at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:272)
2022-09-03T20:18:09.6726600Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-09-03T20:18:09.6726988Z 	at android.os.Handler.$$robo$$android_os_Handler$handleCallback(Handler.java:938)
2022-09-03T20:18:09.6727530Z 	at android.os.Handler.handleCallback(Handler.java)
2022-09-03T20:18:09.6727970Z 	at android.os.Handler.$$robo$$android_os_Handler$dispatchMessage(Handler.java:99)
2022-09-03T20:18:09.6728337Z 	at android.os.Handler.dispatchMessage(Handler.java)
2022-09-03T20:18:09.6728760Z 	at org.robolectric.shadows.ShadowPausedLooper$IdlingRunnable.run(ShadowPausedLooper.java:342)
2022-09-03T20:18:09.6729281Z 	at org.robolectric.shadows.ShadowPausedLooper.executeOnLooper(ShadowPausedLooper.java:271)
2022-09-03T20:18:09.6729948Z 	at org.robolectric.shadows.ShadowPausedLooper.idle(ShadowPausedLooper.java:89)
2022-09-03T20:18:09.6730545Z 	at org.robolectric.android.internal.LocalControlledLooper.drainMainThreadUntilIdle(LocalControlledLooper.java:15)
2022-09-03T20:18:09.6731216Z 	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:318)
2022-09-03T20:18:09.6731618Z 	... 24 more
2022-09-03T20:18:09.6731734Z 
2022-09-03T20:18:09.6731814Z FAILURES!!!
2022-09-03T20:18:09.6732214Z Tests run: 40,  Failures: 1
```

Which should also show when run locally with Bazel or Gradle.

It seems like the failures do have logs--are you referring to a specific check?

JishnuGoyal committed:
Merge remote-tracking branch 'upstream/develop' into modify_chapter_list_design

JishnuGoyal committed:
update test

JishnuGoyal committed:
Merge branch 'develop' into modify_chapter_list_design

JishnuGoyal committed:
update logic

JishnuGoyal committed:
Merge branch 'develop' into modify_chapter_list_design

# Conflicts:
#	app/src/main/res/layout/lessons_chapter_view.xml
#	model/src/main/proto/topic.proto

JishnuGoyal committed:
merge with develop and fix minor bug

JishnuGoyal committed:
these two tests together validate that chapter list is expanded by default

JishnuGoyal committed:
lint

JishnuGoyal committed:
ci fail check

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:
```diff
@@ -212,7 +212,7 @@ class TopicListController @Inject constructor(
         )
       }
-      val firstStoryId = storyRecords.first().storyId
+      val firstStoryId = storyRecords.getOrNull(0)?.storyId
```
Still needed or can this be removed?

JishnuGoyal replied:
yes, this is still needed to set the field "first_topic_id" of the topicsummary proto

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -343,6 +343,21 @@ message TopicSummary {
+  // The old name corresponding to the topic (as a string).
+  reserved 2;
+
+
```
Nit: remove empty line.

JishnuGoyal replied:
done

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -343,6 +343,21 @@ message TopicSummary {
+  reserved 2;
+
+
+  string first_story_id = 9;
```
Is this needed both here & in EphemeralTopicSummary? Also, this is missing a doc string.

JishnuGoyal replied:
this has been included in both. Added a  doc string as well.

BenHenning reviewed (commented):
Thanks @JishnuGoyal. Took another pass. I think one of the past comments hasn't been addressed yet--please make sure to before sending this back.

JishnuGoyal committed:
nits

rt4914 commented on app/src/main/res/drawable/chapter_white_bg_with_bright_green_border.xml:
```diff
@@ -0,0 +1,6 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+  android:shape="rectangle">
+  <stroke android:color="@color/color_def_bright_green" android:width="4dp"/>
+  <solid android:color="@color/color_def_white"/>
```
PLease file and issue mentioning that this new design needs Dark-mode mock too.

JishnuGoyal replied:
done

rt4914 commented on app/src/main/res/layout-sw600dp-land/topic_lessons_story_summary.xml:
```diff
@@ -141,37 +141,37 @@
             android:layout_height="wrap_content"
-            android:layout_marginTop="8dp"
-            android:layout_marginBottom="8dp"
+            android:layout_marginTop="1.5dp"
```
Margin should never be 1.5dp. Ideally it should be a multiple of 4. But if needed you can write it as 2dp.

JishnuGoyal replied:
done

rt4914 reviewed (changes requested):
Do these things:

1. The Chapter name in RTL is left aligned but instead it should be right aligned.
2. Make each chapter item minimum `48dp` height similar to mocks https://xd.adobe.com/view/3dca36c2-5115-419c-b25e-0f10526b077c-6899/screen/558af437-bec4-4719-8d30-291da10bbc9d/specs/
3. Add a video showcasing talkback-output correctly. Either it should have a video or captions on screen. https://developer.android.com/guide/topics/ui/accessibility/testing#optional_talkback_developer_settings && https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide

JishnuGoyal commented:
PTAL @rt4914

JishnuGoyal committed:
add suggested changes

rt4914 reviewed (approved):
LGTM, Thanks.

JishnuGoyal commented:
PTAL @BenHenning

BenHenning reviewed (approved):
Thanks @JishnuGoyal! This LGTM!
